### PR TITLE
Big docs & readme update 

### PR DESCRIPTION
**Pull request by barucden**
Okay, this one's a little bigger.

The goal is to simplify the docs: to contain only what's this package about so we can keep it up-to-date, no duplicates (rather use cross-references), simpler structure, less talking.

The biggest change in README is the addition of contributing section.

Main changes in docs:
 - Menu structure is simpler
 - New example for integration with Flux.jl
 - Added docs for `MapFun` and `AggregateThenMapFun`
 - Removed the "Working with images in Julia" page (instead, it is only a tiny section in introduction that refers to JuliaImages docs)
 - Removed `indices.md`

I plan to do a few more smaller changes in the future (like adding more examples, simplify further, etc.)

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -71,112 +71,90 @@ input in one single pass.
...
+## Working with images in Julia
 
-```@eval
-# Pages = [joinpath("generated", fname) for fname in readdir("generated") if splitext(fname)[2] == ".md"]
-# Depth = 2
-```
+Augmentor exists along other packages in the
+[JuliaImages](https://juliaimages.org/) ecosystem. To learn how images are
+treated in Julia, how pixels are represented, and more, read [the
+documentation](https://juliaimages.org/stable/tutorials/quickstart/).
```
Comment (johnnychen94) [inline docs/src/index.md:157]:
```suggestion
JuliaImages documentation](https://juliaimages.org/stable/tutorials/quickstart/).
```

File: docs/Project.toml [hunk 2]
```diff
@@ -3,12 +3,16 @@ Augmentor = "02898b10-1f73-11ea-317c-6393d7073e15"
 DemoCards = "311a05b2-6137-4a5a-b473-18580a3d38b5"
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
 FileIO = "5789e2e9-d7fb-5bc7-8068-2c6fae9b9549"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+ISICArchive = "0f918735-7648-5aeb-9b51-2c108d137345"
```
Comment (johnnychen94) [inline docs/Project.toml]: I guess this is not used anymore?

```suggestion
```

**Review by johnnychen94 — COMMENTED** [review]
I didn't check every single line but it looks good to me, I also agree some of the contents here are redundant. I kept some of the old contents during the transition because I didn't think of a good replacement for them. As a non-native English speaker, I enjoy @Evizero's words a lot.

File: docs/Project.toml (hunk 2 above)
Comment (barucden) [reply docs/Project.toml]: It is still used in Getting started. I think we can keep it, it looks better when there are more kinds of images, not just MNIST.

**Comment by barucden** [comment]
Yeah, the sentences read nicely. For example, the part about images in Julia was quite educational. I almost felt pity removing it. On the other hand, maintaining the documentation is perhaps harder than maintaining the code, so I aimed to cut down anything that is not necessary (in this specific case, the documentation is available elsewhere -- JuliaImages docs). Also, I think having less text is better for the users as it's easier to skim through.

Edit: Let's keep this PR open for a while so @Evizero has a chance to express his opinion.

**Comment by Evizero** [comment]
Thanks for the kind words. While it was fun writing the images tutorial page, I do think it became outdated quite a while ago. I defer to your combined judgement which has been excellent so far. Thank you for your hard work

File: README.md [hunk 3]
```diff
@@ -7,18 +7,34 @@
...
+
+## Overview
+
+Augmentor.jl provides many augmentation operations such as rotations, flipping,
+blurring, and more. See the
+[documentation](https://evizero.github.io/Augmentor.jl/stable/operations/) for
+the complete list of available operations.
+
+The package exports the `|>` operator, which **composes** operations, forming a
+pipeline.
```
Comment (johnnychen94) [inline README.md]: "export" isn't very accurate here, because by export it is usually a new function provided by this package, but `|>` is owned by the `Base` module.

```suggestion
The package uses the `|>` operator to **compose** operations into a pipeline.
```

File: docs/operations/misc/general.jl [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# ---
+# title: General functions
```
Comment (johnnychen94) [inline docs/operations/misc/general.jl]: How about the title name "Higher order functions"` as it's more understandable in the context of functional programming?

File: docs/src/examples/flux.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
+# Integration with Flux.jl
+
+This example shows a way to use Augmentor to provide images for training
+[Flux.jl](https://github.com/FluxML/Flux.jl/) models. We will be using the
+[MNIST database of handwritten digits](http://yann.lecun.com/exdb/mnist/) as
+our input data.
+
+To skip all the talking and see the code, go ahead to [Complete example](@ref).
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
To skip all the talking and see the code, go ahead to [Complete example](@ref flux_mnist_complete_example).
```

File: docs/src/examples/flux.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+batches = mappedarray(augmentbatch, batchview((X, y), maxsize=batch_size))
+
+nothing # hide
+```
+
+Iterating over batches will now produce augmented images. No other changes are
+required.
+
+## Complete example
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
## [Complete example](@id flux_mnist_complete_example)
```

File: docs/src/examples/flux.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+n_instances = 32
+batch_size = 32
+n_epochs = 16
+
+X = Flux.unsqueeze(MNIST.traintensor(Float32, 1:n_instances), 3)
+y = Flux.onehotbatch(MNIST.trainlabels(1:n_instances), 0:9)
+
+# size(X) == (28, 28, 1, 32)
+# size(y) == (10, 32)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md:36]: I'm not sure if this works, but the idea is to test it during the documentation generation.

```suggestion
# size(X) == (28, 28, 1, 32)
# size(y) == (10, 32)
@assert size(X) == (28, 28, 1, 32) # hide
@assert size(y) == (10, 32) # hide
```

File: docs/src/examples/flux.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+model, pick a loss function and an optimizer, and start the training.
+
+```@example flux
+using Flux, MLDatasets, MLDataUtils
+
+n_instances = 32
+batch_size = 32
+n_epochs = 16
+
+X = Flux.unsqueeze(MNIST.traintensor(Float32, 1:n_instances), 3)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md:32]:
```suggestion
# Flux requires a 4D numerical array in WHCN (width, height, channel, batch)
# format thus we need to insert a dummy dimension to indicate `C=1`(gray image).
X = Flux.unsqueeze(MNIST.traintensor(Float32, 1:n_instances), 3)
```

File: docs/src/examples/flux.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    Flux.train!(loss, params(predict), batches, opt)
+end
+
+nothing # hide
+```
+
+## Adding augmentation
+
+First of all, we remove `Flux.unsqueeze` from the image tensor. This is
+required for Augmentor to correctly process the batches.
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
Augmentor aims to provide generic image augmentation support for any machine
learning framework and not just deep learning. Except for the grayscale images,
Augmentor assumes every image is an array of `Colorant`. Without loss of generality,
we use `Gray` image here so that the same pipeline also applies to `RGB` image.

!!! warning "Use colorant array whenever you can"
    If you pass a 3d numerical array, e.g., of size `(28, 28, 3)` and interpret it as an RGB
    array, you'll almost definitely get an incorrect result from Augmentor. This is because
    Augmentor and the entire JuliaImages ecosystem uses `Array{RGB{Float32}, 2}` to
    represent an `RGB` array. Without any explicit note, `Array{Float32, 3}` will be
    interpreted as a 3d gray image instead of any colorful image. Just think of the color
    specifications like `Lab`, `HSV` and you'll notice the ambiguity here.
```

File: docs/src/examples/flux.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+```@example flux
+X = MNIST.traintensor(Float32, 1:n_instances)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
X = Gray.(MNIST.traintensor(Float32, 1:n_instances))
y = Flux.onehotbatch(MNIST.trainlabels(1:n_instances), 0:9)
```

File: docs/src/examples/flux.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+outbatch(X) = Array{Float32}(undef, (28, 28, 1, nobs(X)))
+# Takes a batch (images and targets) and augments the images
+augmentbatch((X, y)) = (augmentbatch!(outbatch(X), X, pl), y)
+
+nothing # hide
+```
+
+Finally, we wrap the batches with a [mapped
+array](https://github.com/JuliaArrays/MappedArrays.jl/) in order to augment
+each batch.
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
In many deep learning tasks, the augmentation is applied lazily during the data iteration.
For this purpose, we wrap the batches with a [mapped
array](https://github.com/JuliaArrays/MappedArrays.jl/) in order to augment
each batch only when it's required.
```

File: docs/src/examples/flux.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+batch_size = 32
+n_epochs = 16
+
+X = Flux.unsqueeze(MNIST.traintensor(Float32, 1:n_instances), 3)
+y = Flux.onehotbatch(MNIST.trainlabels(1:n_instances), 0:9)
+
+# size(X) == (28, 28, 1, 32)
+# size(y) == (10, 32)
+
+batches = batchview((X, y), maxsize=batch_size)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]: Maybe it's worth mentioning the batch view layout and also `shuffleobs`. (Need to modify all other following parts as well)

```suggestion
# `data = batches[1]` means the first batch input:
#     - `data[1]` is a batch extracted from `X`
#     - `data[2]` is a batch extracted from `Y`
# We also apply `shuffleobs` to get a random batch view.
batches = batchview(shuffleobs((X, y)), maxsize=batch_size)
```

File: docs/src/examples/flux.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+```@example flux
+using Augmentor
+
+pl = ElasticDistortion(6, 6,
+                       sigma=4,
+                       scale=0.3,
+                       iter=3,
+                       border=true) |>
+     Reshape(28, 28, 1)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
pl = ElasticDistortion(6, 6,
                       sigma=4,
                       scale=0.3,
                       iter=3,
                       border=true) |>
     Reshape(28, 28, 1) |>
     SplitChannels()
```

I actually didn't check if this works well on `RGB`, maybe we'll need `SplitChannels(3)` to https://github.com/Evizero/Augmentor.jl/issues/99 make it work.

File: docs/src/examples/flux.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+Finally, we wrap the batches with a [mapped
+array](https://github.com/JuliaArrays/MappedArrays.jl/) in order to augment
+each batch.
+
+```@example flux
+using MappedArrays
+
+batches = mappedarray(augmentbatch, batchview((X, y), maxsize=batch_size))
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
batches = batchview((X, y), maxsize=batch_size)
# lazy evaluation: the augmentation applies lazily until you indexing into `batches`
batches = mappedarray(augmentbatch, batches)
# eager evaluation: the augmentation applies when this line gets executed.
# batches = augmentbatch.(batches)

# The output is already the expected WHCN format
# size(batches[1][1]) == (28, 28, 1, 32)
# size(batches[1][2]) == (10, 32)
@assert size(batches[1][1]) == (28, 28, 1, 32) # hide
@assert size(batches[1][2]) == (10, 32) # hide
```

File: docs/src/examples/flux.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+X = MNIST.traintensor(Float32, 1:n_instances)
+
+nothing # hide
+```
+
+Augmentation is given by an augmentation pipeline. Our pipeline is a
+composition of two operations:
+
+  1. [`ElasticDistortion`](@ref) is the only image operation in this pipeline,
+  2. [`Reshape`](@ref) adds the singleton dimension that is required by Flux.
```
Comment (johnnychen94) [inline docs/src/examples/flux.md]:
```suggestion
Augmentation is given by an augmentation pipeline. Our pipeline is a
composition of three operations:

  1. [`ElasticDistortion`](@ref) is the only image operation in this pipeline,
  2. [`Reshape`](@ref) adds the singleton dimension that is required by Flux.
  3. [`SplitChannels`](@ref) split the colorant array into the plain numerical array so that deep learning frameworks are happy with the layout.
```

File: docs/src/examples/flux.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+     Reshape(28, 28, 1)
+```
+
+Next, we define two helper functions.
+
+```@example flux
+# Creates an output array for augmented images
+outbatch(X) = Array{Float32}(undef, (28, 28, 1, nobs(X)))
+# Takes a batch (images and targets) and augments the images
+augmentbatch((X, y)) = (augmentbatch!(outbatch(X), X, pl), y)
```
Comment (johnnychen94) [inline docs/src/examples/flux.md:120]: Do you think it makes sense to provide a `Label` wrapper for this and unconditionally skips the entire augmentation pipeline? For example,

```julia
augment((X, Label(y)))
```

**Review by johnnychen94 — COMMENTED** [review]
Very well written! The "Label-preserving Transformations" part is very educational; I believe most people are just ignorant about this.

I reckon that we should emphasize the importance of `Colorant` array. Many issues in using this package are due to the old habit of the plain `Array{Float64, 3}` data. I didn't comment on the "Complete example" section so you might need to update it as well.

File: docs/src/examples/flux.md (hunk 14 above)
Comment (barucden) [reply docs/src/examples/flux.md]: I am hesitant about the eager version. The essence of augmentation is to get a bigger and richer dataset. So in each epoch, the batches are augmented over and over (which produces different images when combined with stochastic pipeline).

Comment (johnnychen94) [reply docs/src/examples/flux.md]: I was assuming people never heard about `mappedarray` and I wanted to show them that `mappedarray(auggmentbatch, batches)` is basically equivalent to `augmentbatch.(batches)` except the prior is evaluated lazily.

**Review by johnnychen94 — APPROVED** [review]
@barucden Unless requested, I don't have more comments in. With some final word tweaks, I think we're good to go. It's up to you.

File: docs/src/examples/flux.md (hunk 7 above)
Comment (barucden) [reply docs/src/examples/flux.md:36]: For future reference, this works :)

File: docs/src/examples/flux.md (hunk 14 above)
Comment (barucden) [reply docs/src/examples/flux.md]: Okay, makes sense!

File: docs/src/examples/flux.md (hunk 13 above)
Comment (barucden) [reply docs/src/examples/flux.md]: Ended up doing distortion -> split channels -> permute dimensions.

File: docs/src/examples/flux.md (hunk 16 above)
Comment (barucden) [reply docs/src/examples/flux.md:120]: Good idea. I created #100 so we can discuss there. I think we need to first make `augmentbatch!` work with semantic wrappers.

**Comment by barucden** [comment]
Thanks for the feedback! I am merging it in now. We can continue to improve the documentation iteratively.

File: docs/operations/misc/higherorder.jl [hunk 17]
```diff
@@ -0,0 +1,28 @@
+# ---
+# title: Higher-order functions
+# description: a set of helper opeartions that allow applying any function
+# ---
```
Comment (johnnychen94) [inline docs/operations/misc/higherorder.jl:4]: Looks like we need a cover image here.

Comment (barucden) [reply docs/operations/misc/higherorder.jl:4]: Is it required? I could not think of any since the operations are really general.

Comment (johnnychen94) [reply docs/operations/misc/higherorder.jl:4]: No, but I guess it somehow doesn't look very nice if using the fallback cover... I really have a bad taste in designing the icon/cover. :(

Comment (barucden) [reply docs/operations/misc/higherorder.jl:4]: I will try to add a better cover picture in near future
